File: frontend/src/objects/level/submission.ts
import { z } from "zod";
import { SubmissionStatusSchema } from "../system/system-objects.js";

export const SubmissionSchema = z.object({
  id: z.string().min(1),
  levelId: z.string().min(1),
  submitterId: z.string().min(1),
  status: SubmissionStatusSchema,
  reviewerId: z.string().min(1).optional(),
  reviewNote: z.string().max(1000).optional(),
  submittedAt: z.string(),
  reviewedAt: z.string().optional(),
});

export type Submission = z.infer<typeof SubmissionSchema>;
export { SubmissionWithLevelSchema } from "./submission-with-level.js";
export type { SubmissionWithLevel } from "./submission-with-level.js";

File: backend/microservice/src/level/objects/Submission.scala
package microservice.level.objects

import io.circe.generic.semiauto._
import io.circe.{Decoder, Encoder}
import microservice.system.objects.SubmissionStatus

final case class Submission(
  id: String,
  levelId: String,
  submitterId: String,
  status: SubmissionStatus,
  reviewerId: Option[String],
  reviewNote: Option[String],
  submittedAt: String,
  reviewedAt: Option[String]
)

object Submission {
  implicit val encoder: Encoder[Submission] = deriveEncoder
  implicit val decoder: Decoder[Submission] = deriveDecoder
}
